pkg/quota/admission/clusterresourcequota/accessor.go

```diff
@@ -0,0 +1,161 @@
+package clusterresourcequota
+
+import (
+	"time"
+
+	lru "github.com/hashicorp/golang-lru"
+
+	kapi "k8s.io/kubernetes/pkg/api"
+	kapierrors "k8s.io/kubernetes/pkg/api/errors"
+	utilquota "k8s.io/kubernetes/pkg/quota"
+	"k8s.io/kubernetes/pkg/storage/etcd"
+	utilwait "k8s.io/kubernetes/pkg/util/wait"
+
+	oclient "github.com/openshift/origin/pkg/client"
+	ocache "github.com/openshift/origin/pkg/client/cache"
+	quotaapi "github.com/openshift/origin/pkg/quota/api"
+	"github.com/openshift/origin/pkg/quota/controller/clusterquotamapping"
+)
+
+// QuotaAccessor abstracts the get/set logic from the rest of the Evaluator.  This could be a test stub, a straight passthrough,
+// or most commonly a series of deconflicting caches.
+type QuotaAccessor interface {
+	// UpdateQuotaStatus is called to persist final status.  This method should write to persistent storage.
+	// An error indicates that write didn't complete successfully.
+	UpdateQuotaStatus(newQuota *kapi.ResourceQuota) error
+
+	// GetQuotas gets all possible quotas for a given namespace
+	GetQuotas(namespace string) ([]kapi.ResourceQuota, error)
+}
+
+type quotaAccessor struct {
+	quotaLister     *ocache.IndexerToClusterResourceQuotaLister
+	namespaceLister *ocache.IndexerToNamespaceLister
+	quotaClient     oclient.ClusterResourceQuotasInterface
+
+	quotaMapper clusterquotamapping.ClusterQuotaMapper
+
+	// updatedQuotas holds a cache of quotas that we've updated.  This is used to pull the "really latest" during back to
+	// back quota evaluations that touch the same quota doc.  This only works because we can compare etcd resourceVersions
+	// for the same resource as integers.  Before this change: 22 updates with 12 conflicts.  after this change: 15 updates with 0 conflicts
+	updatedQuotas *lru.Cache
+}
+
+// newQuotaAccessor creates an object that conforms to the QuotaAccessor interface to be used to retrieve quota objects.
+func newQuotaAccessor(quotaLister *ocache.IndexerToClusterResourceQuotaLister, namespaceLister *ocache.IndexerToNamespaceLister, quotaClient oclient.ClusterResourceQuotasInterface, quotaMapper clusterquotamapping.ClusterQuotaMapper) *quotaAccessor {
+	updatedCache, err := lru.New(100)
+	if err != nil {
+		// this should never happen
+		panic(err)
+	}
+
+	return &quotaAccessor{
+		quotaLister:     quotaLister,
+		namespaceLister: namespaceLister,
+		quotaClient:     quotaClient,
+		quotaMapper:     quotaMapper,
+		updatedQuotas:   updatedCache,
+	}
+}
+
+func (e *quotaAccessor) UpdateQuotaStatus(newQuota *kapi.ResourceQuota) error {
+	quota, err := e.quotaLister.Get(newQuota.Name)
+	if err != nil {
+		return err
+	}
+	quota = e.checkCache(quota)
+
+	// make a copy
+	obj, err := kapi.Scheme.Copy(quota)
+	if err != nil {
+		return err
+	}
+	quota = obj.(*quotaapi.ClusterResourceQuota)
```

Comment:
probably should check the cast as well, some folks are more nit picky on that than i am.